File: Recursion/subsequence_pattern/CombinationSum3.py
def _helper(value: int, count: int, target: int,
            result: list[list[int]], current: list[int]) -> None:
    if len(current) == count:
        if target == 0:
            result.append(current.copy())
        return
    for candidate in range(value, 10):
        if candidate > target:
            break
        current.append(candidate)
        _helper(candidate + 1, count, target - candidate, result, current)
        current.pop()


def combination_sum3(count: int, target: int) -> list[list[int]]:
    result = []
    _helper(1, count, target, result, [])
    return result

File: Recursion/subsequence_pattern/combination_sum_3.cpp
#include <vector>

class combination_sum_3 {
    static void helper(int index, int k, int target, std::vector<std::vector<int>>& result,
                       std::vector<int>& current) {
        if (static_cast<int>(current.size()) == k) {
            if (target == 0) result.push_back(current);
            return;
        }
        for (int value = index; value <= 9; ++value) {
            if (value > target) break;
            current.push_back(value);
            helper(value + 1, k, target - value, result, current);
            current.pop_back();
        }
    }
public:
    static std::vector<std::vector<int>> combinationSum3(int k, int n) {
        std::vector<std::vector<int>> result;
        std::vector<int> current;
        helper(1, k, n, result, current);
        return result;
    }
};